<a href="https://colab.research.google.com/github/xthomas/dataviz/blob/master/Klick_Role_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
from google.colab import drive

In [2]:
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [3]:
!ls "/content/gdrive/My Drive/3. Klick/Corporate"

'Klick Role Analysis.ipynb'	   'Reporting Portal - User New Roles.xlsx'
 MaskingTest_MedicalStrategy.xlsx   UnmaskingTest_aFewRolesMore.xlsx


In [4]:
KlickRoles = pd.read_excel ('/content/gdrive/My Drive/3. Klick/Corporate/Reporting Portal - User New Roles.xlsx')

In [5]:
list (KlickRoles)

['UserID', 'Full Name', 'Title', 'RoleID', 'Labor Role', 'CST']

In [6]:
#how many roles at each Client Service Team?
KlickRoles.groupby("CST")["RoleID"].count()

CST
Brave            19
Edge             83
Epic            103
Katalyst          5
Klick Health    387
Klick Inc       123
Nova            231
Quest            82
Rev             109
Sensei           77
Spark           182
Stack             1
Team Takeda     185
Name: RoleID, dtype: int64

In [7]:
KlickRoles.groupby("Labor Role").count()

UserID  Full Name  ...  RoleID  CST
Labor Role                                                        ...             
Activation Strategy                                 4          4  ...       4    4
Activation Strategy, Director                      15         15  ...      15   15
Activation Strategy, Executive Vice President       1          1  ...       1    1
Activation Strategy, Junior                         6          6  ...       6    6
Activation Strategy, Manager                        2          2  ...       2    2
...                                               ...        ...  ...     ...  ...
Web Development                                    66         66  ...      66   66
Web Development, Director                           1          1  ...       1    1
Web Development, Junior                             2          2  ...       2    2
Web Development, Manager                            1          1  ...       1    1
Web Development, Senior                            16         16  ...      16   16

[222 rows x 5 columns]

In [8]:
# Looks like the Role ID prefix can be used to categorize all these roles... 
KlickRoles.RoleID.unique()

array(['TECHMMGD', 'EXCTVMGD', 'USBLTVPR', 'TECHMDIR', 'WEBDVSNR',
       'TECHMVPR', 'FINCESVP', 'DESGNDIR', 'PRMGTDIR', 'TECHMSDR',
       'PRMGTSTD', 'EXCTVSTD', 'PRMGTVPR', 'USBLTDIR', 'SLSTGVPR',
       'TECHMSVP', 'MDSEMVPR', 'PRMGTSVP', 'DESGNMGR', 'FINCEMGR',
       'BUSDVSVP', 'CLSRVMGD', 'EXCTVSVP', 'BUSDVVPR', 'EDITRVPR',
       'CLSRVVPR', 'SYSADSTD', 'SYSADDIR', 'BUSDVSDR', 'TECHMADR',
       'CLSRVSTD', 'PRMGTMGR', 'QLTASVPR', 'CONSTMGD', 'CRMNGADR',
       'CLSRVSDR', 'QLTASMGR', 'CORPTADR', 'APPDVSNR', 'CLSRVSVP',
       'DESGNSTD', 'APPDVMGR', 'ARTDRMGR', 'FINCEVPR', 'EDITRSVP',
       'CLSRVDIR', 'MDRSHVPR', 'CRMNGVPR', 'STRGYEVP', 'SYSADSNR',
       'FINCEDIR', 'PRMGTSDR', 'CRMSTSDR', 'ARTDRSTD', 'EDITRDIR',
       'STRGYSVP', 'CPYWRMGR', 'APPDVSTD', 'CONSTDIR', 'STRGYVPR',
       'OPRTNSTD', 'CORPTSTD', 'BUSANSNR', 'MDSEOVPR', 'ANLTCMGR',
       'CORPTVPR', 'OPRTNDIR', 'CORPTMGR', 'CONSTMGR', 'CORPTDIR',
       'EDITRSTD', 'MDSTGDIR', 'EDITRMGR', 'SLSTGSTD', 'PRDCRS

In [9]:
# but there are a LOT of roles...and looks like some overlap
# for example MD isn't just Media, it's also Medical Strategy
startswithmd = KlickRoles[KlickRoles.RoleID.str.startswith('MD')]
startswithmd.RoleID.unique()

array(['MDSEMVPR', 'MDRSHVPR', 'MDSEOVPR', 'MDSTGDIR', 'MDPLNDIR',
       'MDSEOSTD', 'MDPRGVPR', 'MDTTVSDR', 'MDAOPVPR', 'MDPLNSVP',
       'MDSTGSVP', 'MDPLNVPR', 'MDSEMMGR', 'MDSTGMGD', 'MDSEMDIR',
       'MDSTGVPR', 'MDTTVSTD', 'MDAOPADR', 'MDPLNADR', 'MDSTGSDR',
       'MDPLNSDR', 'MDPLNMGR', 'MDPLNSTD', 'MDAOPMGR', 'MDSTGMGR',
       'MDPLNJNR', 'MDPRGMGR', 'MDPDSMGR', 'MDSTGADR', 'MDAOPJNR',
       'MDTTVADR', 'MDSEOADR', 'MDPDSSTD', 'MDPDSDIR', 'MDPRGADR',
       'MDSEMADR', 'MDRSHSTD', 'MDRSHJNR', 'MDPDSJNR', 'MDTTVJNR',
       'MDSEMSTD', 'MDPDSSDR', 'MDSTGSTD', 'MDPRGSTD', 'MDSEOMGR',
       'MDPDSADR', 'MDPRGJNR', 'MDPRGSDR', 'MDSTGJNR'], dtype=object)

In [10]:
startswithmd.head(10)

UserID  ...           CST
25     4001  ...  Klick Health
65     4601  ...  Klick Health
107    5016  ...  Klick Health
125    5172  ...  Klick Health
138    5305  ...          Nova
177    5640  ...  Klick Health
189    5726  ...  Klick Health
232    5979  ...      Katalyst
238    6016  ...  Klick Health
259    6130  ...   Team Takeda

[10 rows x 6 columns]

In [11]:
startswithmd.shape

(132, 6)

☝ That looks 'scannable' in a spreadsheet

Let's see if we can decode **Departments** by the **RoleID prefix**


First...create 'masks' for the main departments I want to compare

---




In [12]:
# **MEDIA**
mask1 = ((startswithmd.RoleID.str.startswith('MD')) & (~startswithmd.RoleID.str.startswith('MDST')))

# **MEDICAL STRATEGY**
mask2 = startswithmd.RoleID.str.startswith('MDST')

In [13]:
# then write a new CATEGORY column, based on what mask fits them
# let's start with something easier to see, like this "MD" slice
startswithmd['Category'] = np.select([mask1, mask2],['Media','Medical Strategy'],'Confusing')
startswithmd.to_excel('/content/gdrive/My Drive/3. Klick/Corporate/MaskingTest_MedicalStrategy.xlsx')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


A [quick scan](https://drive.google.com/file/d/1tb3GCNiiu4BeFIEg7jcrF2Q9Ur6Ubhju/view?usp=sharing) of **the 132** RoleIDs starting with "MD" looks like those masks work well enough!

Now to create more masks, for those other departments we want to compare, by CST

In [15]:
# **01. MEDIA**
mask1 = ((KlickRoles.RoleID.str.startswith('MD')) & (~startswithmd.RoleID.str.startswith('MDST')))

# **02. STRATEGY**
mask2 = (KlickRoles.RoleID.str.startswith('MDST') | 
         KlickRoles.RoleID.str.startswith('BRST')
         )

# **03. PROJECT MANAGEMENT**
mask3 = KlickRoles.RoleID.str.startswith('PRMG')

# **04. TECHNICAL**
mask4 = (KlickRoles.RoleID.str.startswith('TECH') | 
         KlickRoles.RoleID.str.startswith('APP') | 
         KlickRoles.RoleID.str.startswith('QLTA') | 
         KlickRoles.RoleID.str.startswith('WEBDV') |
         KlickRoles.RoleID.str.startswith('SYSAD') |         
         KlickRoles.RoleID.str.startswith('WEBDV')       
         )

# **05. CLIENT SERVICES**
mask5 = KlickRoles.RoleID.str.startswith('CLSR')

# **06. CORPORATE OVERHEAD**
mask6 = (KlickRoles.RoleID.str.startswith('EXCT') | 
         KlickRoles.RoleID.str.startswith('FINCE') | 
         KlickRoles.RoleID.str.startswith('CORP') | 
         KlickRoles.RoleID.str.startswith('BUSDV') |
         KlickRoles.RoleID.str.startswith('OPRT')
         )


In [16]:
# good enough, write those to a new CATEGORY column, based on what mask fits them
# dump it to a spreadsheet to scan for matches
KlickRoles['Category'] = np.select([mask1, mask2, mask3, mask4, mask5, mask6],[
                                                   'Media',
                                                   'Medical Strategy',
                                                   'Project Management',
                                                   'Technical Support',
                                                   'Client Services',
                                                   'Corporate Overhead'
                                                   ],'Confusing')
KlickRoles.to_excel('/content/gdrive/My Drive/3. Klick/Corporate/UnmaskingTest_aFewRolesMore.xlsx')